### PR TITLE
lapack/gonum: rewrite Dlassq and its test

### PR DESCRIPTION
**Pull request by vladimir-ch**
Please take a look.

Updates #1623 

I'm submitting as draft while I clarify the correct values of `dssml` and `dsbig` with the Reference Lapack.

### CODE REVIEW

File: lapack/gonum/dlassq.go [hunk 1]
```diff
@@ -6,10 +6,9 @@ package gonum
...
-// Dlassq updates a sum of squares in scaled form. The input parameters scale and
-// sumsq represent the current scale and total sum of squares. These values are
-// updated with the information in the first n elements of the vector specified
-// by x and incX.
+// Dlassq updates a sum of squares represented in scaled form. Dlassq returns
+// the values scl and smsq such that
+//  scl^2*smsq = X[0]^2 + ... + X[n-1]^2 + scale^2*sumsq
 //
 // Dlassq is an internal routine. It is exported for testing purposes.
 func (impl Implementation) Dlassq(n int, x []float64, incx int, scale float64, sumsq float64) (scl, smsq float64) {
```
Comment (kortschak) [inline lapack/gonum/dlassq.go:14]: Can you note that the implementation is as provided in the supplemental material of https://doi.org/10.1145/3061665?

File: lapack/gonum/lapack.go [hunk 2]
```diff
@@ -54,4 +54,31 @@ const (
...
+	//    ACM Trans Math Softw 44:1--28
+	//    https://doi.org/10.1145/3061665
+	//  - Blue, James L. (1978)
+	//    A Portable Fortran Program to Find the Euclidean Norm of a Vector
+	//    ACM Trans Math Softw 4:15--23
+	//    https://doi.org/10.1145/355769.355771
+	dtsml = 0x1p-511
+	dtbig = 0x1p486
+	dssml = 0x1p537
+	dsbig = 0x1p-538
```
Comment (kortschak) [inline lapack/gonum/lapack.go:83]: I'm happy with these (they match the [reference implementation](https://dl.acm.org/action/downloadSupplement?doi=10.1145%2F3061665&file=978.zip)).

File: lapack/gonum/dlassq.go [hunk 3]
```diff
@@ -22,20 +21,90 @@ func (impl Implementation) Dlassq(n int, x []float64, incx int, scale float64, s
...
+	//  - values smaller than dtsml are scaled up by dssml
+	var (
+		isBig            bool
+		asml, amed, abig float64
+	)
+	for i, ix := 0, 0; i < n; i++ {
+		ax := math.Abs(x[ix])
+		switch {
+		case ax > dtbig:
+			abig += (ax * dsbig) * (ax * dsbig)
```
Comment (kortschak) [inline lapack/gonum/dlassq.go:55]:
```
ax *= dsbig
abig += ax*ax
```
? (if yes, similar below)

The compiler generates the same code either way.

File: lapack/gonum/dlassq.go (hunk 1 above)
Comment (vladimir-ch) [reply lapack/gonum/dlassq.go:14]: Done.

File: lapack/gonum/dlassq.go (hunk 3 above)
Comment (vladimir-ch) [reply lapack/gonum/dlassq.go]: I'm fine with your suggestion. Done at all 4 places.

File: lapack/gonum/lapack.go (hunk 2 above)
Comment (vladimir-ch) [reply lapack/gonum/lapack.go:83]: Ok, I'll remove the draft status.

**Review by kortschak — APPROVED** [review]

**Pull request merged**